dropshot/src/extractor/common.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright 2022 Oxide Computer Company
+
+// XXX-dap TODO-cleanup should the metadata into a separate, shared trait?
+
+use crate::api_description::ApiEndpointParameter;
+use crate::api_description::{ApiEndpointBodyContentType, ExtensionMode};
+use crate::error::HttpError;
+use crate::server::ServerContext;
+use crate::RequestContext;
+
+use async_trait::async_trait;
+
+/// Metadata associated with an extractor including parameters and whether or not
+/// the associated endpoint is paginated.
+pub struct ExtractorMetadata {
+    pub extension_mode: ExtensionMode,
+    pub parameters: Vec<ApiEndpointParameter>,
+}
+
+/// Extractors that require exclusive access to the underyling `hyper::Request`
+///
+/// These extractors usually need to read the body of the request or else modify
+/// how the server treats the rest of it (e.g., websocket upgrade).  There may
+/// be at most one of these associated with any request.
+#[async_trait]
+pub trait ExclusiveExtractor: Send + Sync + Sized {
+    /// Construct an instance of this type from a `RequestContext`.
+    async fn from_request<Context: ServerContext>(
+        rqctx: &RequestContext<Context>,
```

Comment:
Counter-intuitively, this works with just `&RequestContext<Context>`.  That's because what exclusive extractors need mutable access to is the `hyper::Request`, which is in the `RequestContext` but wrapped with `Arc<Mutex<...>>`.  So you can take the lock to access a `&mut hyper::Request`.

After #558, the `hyper::Request` won't even be in the `RequestContext` any more.

So given that it's not necessary to have a `&mut RequestContext` and that the caller is going to pass the same `RequestContext` to the request handler function after calling this one, I prefer giving the extractor just a `&RequestContext`.  Hopefully that all makes sense.